grove/ising/ising_qaoa.py

```diff
@@ -0,0 +1,142 @@
+
+"""
+Finding the minimum energy for an Ising problem by QAOA.
+"""
+import pyquil.api as api
+from grove.pyqaoa.qaoa import QAOA
+from pyquil.paulis import PauliSum, PauliTerm
+from scipy.optimize import minimize
+import numpy as np
+
+CXN = api.SyncConnection()
+
+
+def energy_value(h, J, sol):
+    """
+    Obtain energy of an Ising solution for a given Ising problem (h,J).
+
+    :param h: External magnectic term of the Ising problem. List.
+    :param J: Interaction term of the Ising problem. Dictionary.
+    :param sol: Ising solution. List.
+
+    """
+    ener_ising = 0
+    for elm in J.keys():
+        if elm[0] == elm[1]:
+            ener_ising += J[elm] * int(sol[elm[0]])
+        else:
+            ener_ising += J[elm] * int(sol[elm[0]]) * int(sol[elm[1]])
+    for i in range(len(h)):
+        ener_ising += h[i] * int(sol[i])
+    return ener_ising
+
+
+def print_fun(x):
+    print(x)
+
+
+def ising_trans(x):
+    # Transformation to Ising notation
+    if x == 1:
+        return -1
+    else:
+        return 1
+
+
+def ising_qaoa(h, J, steps=1, rand_seed=None, connection=None, samples=None,
+               initial_beta=None, initial_gamma=None, minimizer_kwargs=None,
+               vqe_option=None):
+    """
+    Ising set up method
+
+    :param h: External magnectic term of the Ising problem. List.
+    :param J: Interaction term of the Ising problem. Dictionary.
+    :param steps: (Optional. Default=1) Trotterization order for the
+                  QAOA algorithm.
+    :param rand_seed: (Optional. Default=None) random seed when beta and
+                      gamma angles are not provided.
+    :param connection: (Optional) connection to the QVM. Default is None.
+    :param samples: (Optional. Default=None) VQE option. Number of samples
+                    (circuit preparation and measurement) to use in operator
+                    averaging.
+    :param initial_beta: (Optional. Default=None) Initial guess for beta
+                         parameters.
+    :param initial_gamma: (Optional. Default=None) Initial guess for gamma
+                          parameters.
+    :param minimizer_kwargs: (Optional. Default=None). Minimizer optional
+                             arguments.  If None set to
+                             {'method': 'Nelder-Mead',
+                             'options': {'ftol': 1.0e-2, 'xtol': 1.0e-2,
+                                        'disp': False}
+    :param vqe_option: (Optional. Default=None). VQE optional
+                             arguments.  If None set to
+                       vqe_option = {'disp': print_fun, 'return_all': True,
+                       'samples': samples}
+
+    """
+
+    cost_operators = []
```

Comment:
Is there a reason you separated out the calculation of the cost and driver and the subsequent QAOA call from the initialization method `ising` below? It seems to me, that those could be combined?